### PR TITLE
feat(redis): support redis failover (sentinel)

### LINKED ISSUE #37: Redis Cluster/Sentinel
From what I can tell from the upstream library, you need to instantiate a client with NewClusterClient/NewSentinelClient in order to setup connections to these types, I've not confirmed this. But was wondering if you either knew before I worked on supporting something like this, and if you had a preference for how we would set it up.

I think realistically it would be better to make it their own providers since the options are vastly different for [Sentinel](https://github.com/go-redis/redis/blob/master/sentinel.go#L21-L70) and for [Cluster](https://github.com/go-redis/redis/blob/master/cluster.go#L26-L80) configurations.

### PR DESCRIPTION
**Pull request by james-d-elliott**
This adds the go-redis/redis FailoverClient which is used for Redis Sentinel. It's mostly a copy of https://github.com/fasthttp/session/tree/master/providers/redis with the necessary tweaks for Sentinel. I am unsure if the SentinelAddrs are required, but I believe they are optional (obtained from the master automatically if not provided).

If merged closes #37.

### CODE REVIEW

**Comment by james-d-elliott** [comment]
I'm going to test this by implementing it in Authelia before I'm happy with this being merged unless you wish to test it. Will move it  from a draft once that's done. You're welcome to give feedback before then if you have any.

**Comment by james-d-elliott** [comment]
So I've tested this with Authelia in my k8s cluster with a fresh Redis Sentinel setup. It works as intended. I ended up going with `redisfailover` as the package name as this is the sentinel provider that seemed to closest match the functionality in `redis`, and instead of calling it `redissentinel` I chose that name since it matches the upstream type, and there is an alternate one for sentinel.

My impression is the failover client is intended for the easiest implementation, and you use the sentinel clientif you want to manually tune behavior and I think it has to be utilized along side the standard redis client. The failover client bundles all of this.

I did experiment with Redis Cluster which is more geared for large data sets. For starters this doesn't seem like a logical use case as session storage should theoretically be pretty small. Second of all the way it works is multi-node writes and sharding. The issue is you have to configure their hash slot concept to get data consistency.

**Comment by savsgio** [comment]
Hi @james-d-elliott,

I have been looking the redis library, and i could see, it's not necessary to create a separated package to 'failover' and 'cluster' client, because all of them implement redis.Cmdable, so into redis package you could do something like this:

```go
type Provider struct {
	cfg Config
	db  redis.Cmdable // Before ==> *redis.Client
}

func New(cfg Config) (*Provider, error) {
	// ....

	db := redis.NewClient(&redis.Options{
		// ...
	})

	// ....

	p := &Provider{
		// ...
		db: db,
	}

	return p, nil
}

func NewFailover(cfg Config) (*Provider, error) {
	// ....

	db := redis.NewFailoverClient(&redis.FailoverOptions{
		// ...
	})

	// ....

	p := &Provider{
		// ...
		db: db,
	}

	return p, nil
}

func NewCluster(cfg Config) (*Provider, error) {
	// ....

	db := redis.NewClusterClient(&redis.ClusterOptions{
		// ...
	})

	// ....

	p := &Provider{
		// ...
		db: db,
	}

	return p, nil
}
```

**Comment by james-d-elliott** [comment]
I'm quite happy to do that. I'll refactor today. @savsgio do you prefer a Config struct with all options, or one for each, or one for each plus a common struct i.e.:

```go
type CommonConfig struct {
  // all common values (between std/failover/cluster)
}

type Config struct {
  CommonConfig
  DB int
  //etc
}

type FailoverConfig struct {
  CommonConfig
  DB int
  SentinelAddrs []string
  //etc
}

type ClusterConfig struct {
  CommonConfig
  // NOT VALID for cluster DB int
  //etc
}
```

**Comment by james-d-elliott** [comment]
Actually thinking about this more. This change would be a breaking one. It's better for me to either expressly setup the structs, or include all options in every config. I will have an example of the later up in the maintenance repo we're using in a few minutes.

**Comment by james-d-elliott** [comment]
I've done some preliminary testing with this and bitnami's redis containers which can be seen in Authelia's Buildkite (HighAvailabiltiy suite):

https://buildkite.com/authelia/authelia/builds/4475#19172e30-698c-42c0-a44c-ad2865524159

**Comment by james-d-elliott** [comment]
As a small aside, I have only tested the NewFailover, I played with testing NewCluster previously but it requires messing with the ClusterSlots command to get the hash slots to work properly otherwise the client doesn't know where to obtain data, and I can't pretend to understand it as it would probably take significant research.

I'm relatively confident what is being exposed to the users is enough with my latest commit. If you had a specific section of code you wanted me to test, I'd be able to test it against my Redis Cluster instance in k8s. Or if you'd like me to remove the cluster code until such a time as I can test it I'd be willing to research it in the next month.

**Comment by savsgio** [comment]
Hi @james-d-elliott,

LGTM the current change, so meanwhile just add the failover client and the cluster client add it later, when you could test it.

File: providers/redis/types.go [hunk 1]
```diff
@@ -90,8 +91,196 @@ type Config struct {
...
+	Password string
+
+	// Database to be selected after connecting to the server.
+	DB int
+
+	// The sentinel master name.
+	MasterName string
+
+	// The sentinel nodes seed list (host:port).
+	Addrs []string
```
Comment (savsgio) [inline providers/redis/types.go:113]: Could you rename to `SentinelAddrs`, please?

Comment (james-d-elliott) [reply providers/redis/types.go:113]: Sure, will do it now.

File: providers/redis/provider.go [hunk 2]
```diff
@@ -43,16 +43,97 @@ func New(cfg Config) (*Provider, error) {
...
+
+// NewFailover returns a new configured redis failover (sentinel) provider
+func NewFailover(cfg FailoverConfig) (*Provider, error) {
+	if cfg.MasterName == "" {
+		return nil, errConfigMasterNameEmpty
+	}
+
+	db := redis.NewFailoverClient(&redis.FailoverOptions{
+		MasterName:         cfg.MasterName,
+		SentinelAddrs:      cfg.Addrs,
```
Comment (savsgio) [inline providers/redis/provider.go:61]: Change it here too, please

Comment (james-d-elliott) [reply providers/redis/provider.go:61]: Oh that's embarrassing...

**Comment by james-d-elliott** [comment]
> Hi @james-d-elliott,
> 
> LGTM the current change, so meanwhile just add the failover client and the cluster client add it later, when you could test it.

Oh didn't see you wanted the cluster client removed, can do that in a bit.

**Comment by james-d-elliott** [comment]
> 
> 
> Hi @james-d-elliott,
> 
> LGTM the current change, so meanwhile just add the failover client and the cluster client add it later, when you could test it.

The naming of the methods in go-redis are rather confusing. There is actually a NewFailoverClient and a NewFailoverClusterClient, I would have assumed the second would be related to Redis Cluster, however it is not from what I can tell in the code.

Based on some more intense testing it looks like NewFailoverClusterClient is a Sentinel client that connects to a "cluster" of Sentinels which allows for a Sentinel failure to not interrupt connectivity, whereas the NewFailoverClient is less resilient (doesn't seem to connect to an alternate Sentinel in a failure, or if it does, it's highly delayed).

**Comment by james-d-elliott** [comment]
@savsgio thoughts on the naming of the newup functions? Or is this ready to merge?

**Review by savsgio — APPROVED** [review]

**Comment by savsgio** [comment]
@james-d-elliott  Thank you so much!

Sorry for delayed, i'm pretty busy lately. 😉

**Comment by james-d-elliott** [comment]
No worries at all I just merged my changes upstream with the maint fork. Will switch back today.

**Pull request merged**